net: ip: igmp: add IGMPv3 support

IVandeVeire opened the pull request:
Added IGMPv3 support by implementing the basics of RFC 3376. I used the already existing structure of IGMPv2 as reference ~~without changing the existing IGMP api~~. IGMPv3 can be enabled by  enabling `CONFIG_NET_IPV4_IGMPV3`.

IVandeVeire committed:
net: ip: utils: changed input arguments of igmp_checksum to net_pkt

Added igmpv3 checksum function to make it possible to calculate the
checksum of a complete igmpv3 pkt at once.

Signed-off-by: Ibe Van de Veire <ibe.vandeveire@basalte.be>

IVandeVeire committed:
net: ip: igmp: add igmpv3 support

Added igmpv3 support based on the already existing structure for igmpv2.
The already existing api is not modified to prevent breaking exisiting
applications.

Signed-off-by: Ibe Van de Veire <ibe.vandeveire@basalte.be>

jukkar commented on subsys/net/ip/igmp.c:
```diff
@@ -65,6 +73,103 @@ static int igmp_v2_create(struct net_pkt *pkt, const struct in_addr *addr,
+	if (!igmp) {
+		return -ENOBUFS;
+	}
+
```
please remove extra empty line

jukkar commented on subsys/net/ip/igmp.c:
```diff
@@ -65,6 +73,103 @@ static int igmp_v2_create(struct net_pkt *pkt, const struct in_addr *addr,
+		return -ENOBUFS;
+	}
+
+	NET_PKT_DATA_ACCESS_DEFINE(group_record_access, struct net_ipv4_igmp_v3_group_record);
+	struct net_ipv4_igmp_v3_group_record *group_record;
```
Please introduce all variables at the beginning of the function

jukkar commented on subsys/net/ip/igmp.c:
```diff
@@ -114,13 +243,15 @@ static int igmp_send(struct net_pkt *pkt)
 	return 0;
 }
 
+#if !defined(CONFIG_NET_IPV4_IGMPV3)
```
Could / should we support both v2 and v3 at the same time?

IVandeVeire replied:
Hmm indeed, we should be able to respond to queries from older versions. I will take a look!

IVandeVeire replied:
I added the detection of the IGMP version of a received query. Now, IGMPv2 queries are also answered when IGMPv3 is enabled.

IVandeVeire commented on include/zephyr/net/igmp.h:
```diff
@@ -49,7 +49,36 @@ static inline int net_ipv4_igmp_join(struct net_if *iface,
+ * @return Return 0 if joining was done, <0 otherwise.
+ */
+#if defined(CONFIG_NET_IPV4_IGMPV3)
+int net_ipv4_igmpv3_join(struct net_if *iface, const struct in_addr *addr, bool include,
```
I created a separate API because IGMPv3 has some additional arguments/functionality. How would you add the filter groups to a request if the API remains unchanged? For leaving a group, Kconfig should do the job indeed.

rlubos replied:
I'm not saying the API has to remain unchanged, I just put under discussion whether it makes sense to have a separate API for IGMPv3 only. Right now it's a bit counter-intuitive for me, that even after enabling IGMPv3 most (if not all) of the networking components would still send IGMPv2 reports. 

Just as an idea, `net_ipv4_igmp_join()` could accept an additional (optional) `struct igmp_param *` argument, with content dependent on the enabled IGMP version, with some reasonable fallback behavior if not provided (no filters for IGMPv3 if that makes sense). Net components that do not care about this additional functionality (is it just filtering or more?) could just skip the parameter.

IVandeVeire replied:
IGMPv3 does only support filters. I try your suggestion and see how that turns out.

jukkar replied:
I agree with @rlubos, the public APIs should not need to mention v2/v3 etc. but be generic as the caller does not need to care what IGMP version we are using.

IVandeVeire replied:
I made the join/leave more generic. Now, the IGMP version is selected by the config. I added an optional param struct like @rlubos suggested.

IVandeVeire replied:
I will add extra tests and needed documentation for the change of the public API once the changes are on point.

rlubos commented on subsys/net/ip/igmp.c:
```diff
@@ -314,6 +569,58 @@ int net_ipv4_igmp_join(struct net_if *iface, const struct in_addr *addr)
+		net_ipaddr_copy(&maddr->sources[i].in_addr.s_addr, &source_list[i].s_addr);
+	}
+
+	maddr->record_type = IGMPV3_CHANGE_TO_EXCLUDE_MODE;
```
`record_type` was just set above.

rlubos commented on subsys/net/ip/igmp.c:
```diff
@@ -314,6 +569,58 @@ int net_ipv4_igmp_join(struct net_if *iface, const struct in_addr *addr)
+		return ret;
+	}
+
+	maddr->record_type =
```
Ditto.

rlubos commented on subsys/net/ip/igmp.c:
```diff
@@ -65,6 +73,103 @@ static int igmp_v2_create(struct net_pkt *pkt, const struct in_addr *addr,
+	/* Setting initial value of chksum to 0 to calculate chksum as described in RFC 3376
+	 * ch 4.1.2
+	 */
+	igmp->chksum = 0;
```
You should clear the checksum before `net_pkt_set_data(pkt, &igmp_access)` above to ensure it's acutally cleared on a packet. Since you're not requesting contiguous access to the header, you may write to a copy of the header here.

jukkar commented on subsys/net/ip/utils.c:
```diff
@@ -711,6 +711,17 @@ uint16_t net_calc_chksum_igmp(uint8_t *data, size_t len)
 #endif /* CONFIG_NET_IPV4_IGMP */
 
+#if defined(CONFIG_NET_IPV4_IGMPV3)
+uint16_t net_calc_chksum_igmpv3(struct net_pkt *pkt)
```
Can't we have just one `net_calc_chksum_igmp()` function that could handle all versions?

IVandeVeire replied:
The `net_calc_chksum_igmp()` is now used for both IGMPv2 and IGMPv3

IVandeVeire commented on subsys/net/ip/net_private.h:
```diff
@@ -238,6 +238,12 @@ enum net_verdict net_ipv4_igmp_input(struct net_pkt *pkt,
 #define net_calc_chksum_igmp(data, len) 0U
```
What do you mean? At the moment, I did not change anything to this prototype?

jukkar replied:
There seem to be two `net_calc_chksum_igmp()` functions, one for v2 and the other one for v3. It is quite confusing and would be better if only one checksum function is used.

IVandeVeire commented:
I will add extra tests and needed documentation for the change of the public API once the changes are on point.

rlubos commented on include/zephyr/net/igmp.h:
```diff
@@ -27,29 +27,37 @@
  */
 #if defined(CONFIG_NET_IPV4_IGMP)
-int net_ipv4_igmp_join(struct net_if *iface, const struct in_addr *addr);
+int net_ipv4_igmp_join(struct net_if *iface, const struct in_addr *addr, struct igmp_param *param);
```
The `param` pointer could also be const.

rlubos commented on include/zephyr/net/igmp.h:
```diff
@@ -27,29 +27,37 @@
 #endif
 
 /**
- * @brief Leave a given multicast group.
```
Is this intended? Seems odd now.

rlubos commented on include/zephyr/net/net_if.h:
```diff
@@ -101,6 +101,17 @@ struct net_if_mcast_addr {
 	uint8_t _unused : 6;
+
+#if defined(CONFIG_NET_IPV4_IGMPV3)
+	/** Filter mode (used in IGMPV3)*/
```
nit: Please add an empty space before `*/`, below as well.

rlubos commented on subsys/net/ip/igmp.c:
```diff
@@ -56,14 +70,119 @@ static int igmp_v2_create(struct net_pkt *pkt, const struct in_addr *addr,
+		return -ENOBUFS;
+	}
+
+	for (int i = 0; i < NET_IF_MAX_IPV4_MADDR; i++) {
```
Shouldn't this iterate until `i < mcast_len` instead? `igmpv3_send_generic()` will only provide a single address to put into the report.

IVandeVeire replied:
Indeed

rlubos reviewed (commented):
Thank you for addressing the feedback so far. Spotted a few more nits, but otherwise looks pretty good to me.

jukkar commented on subsys/net/ip/igmp.c:
```diff
@@ -121,6 +264,7 @@ static int send_igmp_report(struct net_if *iface,
 	struct net_pkt *pkt = NULL;
 	int i, count = 0;
 	int ret = 0;
+	const struct in_addr all_systems = ALL_SYSTEMS_ADDR;
```
here too, `static const ...`

jukkar commented on subsys/net/ip/igmp.c:
```diff
@@ -195,41 +339,159 @@ static int send_igmp_report(struct net_if *iface,
+	struct net_if_ipv4 *ipv4 = iface->config.ip.ipv4;
+	struct net_pkt *pkt = NULL;
+	int i, group_count = 0, source_count = 0;
+	const struct in_addr all_systems = ALL_SYSTEMS_ADDR;
+	const struct in_addr igmp_multicast_addr = IGMP_MULTICAST_ADDR;
```
Ditto. Perhaps these consts could be even placed at the start of the file instead of placing them in each function.

jukkar commented on subsys/net/ip/igmp.c:
```diff
@@ -323,28 +660,35 @@ int net_ipv4_igmp_leave(struct net_if *iface, const struct in_addr *addr)
 void net_ipv4_igmp_init(struct net_if *iface)
 {
 	struct net_if_mcast_addr *maddr;
+	const struct in_addr all_systems = ALL_SYSTEMS_ADDR;
```
ditto

jukkar reviewed (changes requested):
Looks good, I found couple of minor nits that we could change.

IVandeVeire commented on include/zephyr/net/net_if.h:
```diff
@@ -101,6 +101,17 @@ struct net_if_mcast_addr {
 
 	uint8_t _unused : 6;
+
+#if defined(CONFIG_NET_IPV4_IGMPV3)
+	/** Filter mode (used in IGMPV3) */
+	uint8_t record_type;
+
+	/** Number of sources to be used by the filter */
+	uint16_t sources_len;
+
+	/** Sources to filter on */
+	struct net_addr sources[CONFIG_NET_IF_MCAST_IPV4_SOURCE_COUNT];
+#endif
```
I don't really get how there can be a hole now?
I think the current layout is like this?

|                   32 bit net_addr                  |
|1 bit | 1 bit | 6 bit | 8 bit |16 bit src len |
|                   32 bit net_addr                  |
                             ...

Or is there any reason to order it on descending length?

IVandeVeire committed:
doc: migration guide: add note about IGMP migration

Added note about the migration steps needed to support the new IGMP api.
The api now expects an additional argument used for joining an IGMPv3
group.

Signed-off-by: Ibe Van de Veire <ibe.vandeveire@basalte.be>

jukkar commented on subsys/net/ip/igmp.c:
```diff
@@ -195,41 +339,156 @@ static int send_igmp_report(struct net_if *iface,
+		/* IGMPv3 detected */
+		version = IGMPV3;
+#if !defined(CONFIG_NET_IPV4_IGMPV3)
+		NET_DBG("DROP: IGMPv3 msg received but IGMPv3 support is disabled");
```
We could save couple of bytes with this
```
NET_DBG("DROP: %s msg received but %s support is disabled", "IGMPv3");
```

jukkar commented on subsys/net/ip/igmp.c:
```diff
@@ -195,41 +339,156 @@ static int send_igmp_report(struct net_if *iface,
+	if (version == IGMPV3) {
+		igmpv3_hdr = (struct net_ipv4_igmp_v3_query *)net_pkt_get_data(pkt, &igmpv3_access);
+		if (!igmpv3_hdr) {
+			NET_DBG("DROP: NULL IGMP header");
```
And here
```
NET_DBG("DROP: NULL %s header", "IGMP");
```

jukkar commented on subsys/net/ip/igmp.c:
```diff
@@ -195,41 +339,156 @@ static int send_igmp_report(struct net_if *iface,
+#endif
+		igmpv2_hdr = (struct net_ipv4_igmp_v2_query *)net_pkt_get_data(pkt, &igmpv2_access);
+		if (!igmpv2_hdr) {
+			NET_DBG("DROP: NULL IGMP header");
```
Ditto
And here
```
NET_DBG("DROP: NULL %s header", "IGMP");
```

jukkar reviewed (approved):
@IVandeVeire Thanks for the support, much appreciated!

rlubos reviewed (approved):
LGTM
